### PR TITLE
Fadg - Forward AD native MFEM class with unit tests and example

### PR DESCRIPTION
**Pull request by bslazarov**
This pull request consists of a simple automatic differentiation template class and several unit tests. It is the first of several pull-requests for adding new features to accommodate AD-related implementations in MFEM. The class is templated in fdual.hpp and the unit tests are implemented in tests/unit/linalg/test_fdual.cpp. The implementation is based on dual numbers. 

Two examples - one serial ex71 and one parallel ex71p demonstrate the utilization of AD. 

The native AD implementation can be replaced with external ones. In this case the user has to compile either ADEPT (compile it with flag --dissable-openmp) or just download FADBAD++. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1612](https://github.com/mfem/mfem/pull/1612) | @bslazarov | @tzanio | @v-dobrev + @jandrej + @jamiebramwell | 07/25/20 | ⌛due 03/19/21 | ⌛due 03/26/21 | |

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -147,6 +147,13 @@ examples/hiop/ex9-mesh.*
 examples/hiop/ex9-init.*
 examples/hiop/ex9-final.*
 
+examples/ex71
+examples/pLap/
+examples/ex71p
+examples/pLap/*
+examples/pLap
+examples/pLap/
```
Comment (jandrej) [inline .gitignore:155]: File naming is not in convention with other examples

File: CMakeLists.txt [hunk 2]
```diff
@@ -296,6 +296,19 @@ if (MFEM_USE_HIOP)
...
+        # find_package updates ADEPT_FOUND, ADEPT_INCLUDE_DIRS, ADEPT_LIBRARIES
+endif()
+
+# FADBAD++ package
+if (MFEM_USE_FADBADPP)
+	find_package(FADBADPP REQUIRED)
+        # find_package updates FADBADPP_FOUND, FADBADPP_INCLUDE_DIRS, FADBADPP_LIBRARIES
+endif()
+
+
```
Comment (jandrej) [inline CMakeLists.txt:311]: Extra empty line

File: CMakeLists.txt [hunk 3]
```diff
@@ -357,6 +370,7 @@ endif()
 #    be before SuiteSparse.
 set(MFEM_TPLS MPI_CXX OPENMP BLAS LAPACK METIS HYPRE SuiteSparse SUNDIALS PETSC
     SLEPC MESQUITE SuperLUDist STRUMPACK AXOM CONDUIT Ginkgo GNUTLS GSLIB NETCDF
+    ADEPT FADBADPP
     MPFR PUMI HIOP POSIXCLOCKS MFEMBacktrace ZLIB OCCA CEED RAJA UMPIRE ADIOS2)
```
Comment (jandrej) [inline CMakeLists.txt:374]: formatting

File: INSTALL [hunk 4]
```diff
@@ -448,6 +448,19 @@ MFEM_USE_HIOP = YES/NO
   Enable the usage of HiOp (https://github.com/LLNL/hiop) in MFEM. HiOp is an
   HPC solver for nonlinear optimization problems.
 
+MFEM_USE_ADEPT = YES/NO
+  Enable automatic differentiation using ADEPT library.
```
Comment (jandrej) [inline INSTALL:452]: `using the ADEPT library`

File: INSTALL [hunk 5]
```diff
@@ -448,6 +448,19 @@ MFEM_USE_HIOP = YES/NO
   Enable the usage of HiOp (https://github.com/LLNL/hiop) in MFEM. HiOp is an
   HPC solver for nonlinear optimization problems.
 
+MFEM_USE_ADEPT = YES/NO
+  Enable automatic differentiation using ADEPT library. 
+  (http://www.met.reading.ac.uk/clouds/adept) 
+  Please, compile the library with  flag --disable-openmp.
```
Comment (jandrej) [inline INSTALL:454]: double space `with  flag`

File: INSTALL [hunk 6]
```diff
@@ -448,6 +448,19 @@ MFEM_USE_HIOP = YES/NO
...
+
+MFEM_USE_FADBADPP = YES/NO
+  Enable automatic differentiation using FADBAD++ library.
```
Comment (jandrej) [inline INSTALL:457]: `using the FADBAD++ library`

File: INSTALL [hunk 7]
```diff
@@ -635,6 +648,16 @@ The specific libraries and their options are:
   URL: https://github.com/LLNL/hiop
   Options: HIOP_OPT, HIOP_LIB.
 
+- ADEPT (optional), used with MFEM_USE_ADEPT = YES
+  URL: www.met.reading.ac.uk/clouds/adept/
+  Options: ADEPT_OPT, ADEPT_LIB
+  Versions: 1.1 and 2.0.5
+
+- FADBAD++ (optiobal) , used with MFEM_USE_FADBADPP = YES
```
Comment (jandrej) [inline INSTALL:656]: s/optiobal/optional and extra space before comma

File: config/cmake/MFEMConfig.cmake.in [hunk 8]
```diff
@@ -50,6 +50,9 @@ set(MFEM_USE_CEED @MFEM_USE_CEED@)
 set(MFEM_USE_UMPIRE @MFEM_USE_UMPIRE@)
 set(MFEM_USE_SIMD @MFEM_USE_SIMD@)
 set(MFEM_USE_ADIOS2 @MFEM_USE_ADIOS2@)
+set(MFEM_USE_ADEPT  @MFEM_USE_ADEPT@)
```
Comment (jandrej) [inline config/cmake/MFEMConfig.cmake.in:53]: double space

File: config/cmake/config.hpp.in [hunk 9]
```diff
@@ -156,4 +156,14 @@
...
+// use ADEPT library for AD
+#cmakedefine MFEM_USE_ADEPT
+
+// use FADBAD++ library for AD
+#cmakedefine MFEM_USE_FADBADPP
+
+// use forward mode for automatic differentiation
+#cmakedefine MFEM_USE_ADFORWARD
+
+
```
Comment (jandrej) [inline config/cmake/config.hpp.in:168]: extra empty line

File: config/config.hpp.in [hunk 10]
```diff
@@ -163,4 +163,14 @@
...
+// use ADEPT library for AD
+// #define MFEM_USE_ADEPT
+
+// use FADBAD++ library for AD
+// #define MFEM_USE_FADBADPP
+
+// use forward mode for automatic differentiation
+// #define MFEM_USE_ADFORWARD
+
+
```
Comment (jandrej) [inline config/config.hpp.in:175]: extra empty line

File: config/defaults.cmake [hunk 11]
```diff
@@ -52,6 +52,9 @@ option(MFEM_USE_CEED "Enable CEED" OFF)
 option(MFEM_USE_UMPIRE "Enable Umpire" OFF)
 option(MFEM_USE_SIMD "Enable use of SIMD intrinsics" ON)
 option(MFEM_USE_ADIOS2 "Enable ADIOS2" OFF)
+option(MFEM_USE_ADEPT "Enable AD using ADEPT"   OFF)
```
Comment (jandrej) [inline config/defaults.cmake:55]: extra whitespaces

File: config/defaults.cmake [hunk 12]
```diff
@@ -190,6 +193,13 @@ set(BLAS_LIBRARIES "" CACHE STRING "The BLAS library.")
 set(LAPACK_INCLUDE_DIRS "" CACHE STRING "Path to LAPACK headers.")
 set(LAPACK_LIBRARIES "" CACHE STRING "The LAPACK library.")
 
+set(ADEPT_INCLUDE_DIRS "${MFEM_DIR}/../adept-1.1/include" CACHE STRING "Path to ADEPT headers.")
+set(ADEPT_LIBRARIES    "-L${MFEM_DIR}/../adept-1.1/lib -ladept" CACHE STRING "The ADEPT library.")
+
+set(FADBADPP_INCLUDE_DIRS "${MFEM_DIR}/../FADBAD++" CACHE STRING "Path to FADBAD++ headers.")
+set(FADBADPP_LIBRARIES    "")
```
Comment (jandrej) [inline config/defaults.cmake:200]: alignment looks off

File: config/defaults.cmake [hunk 13]
```diff
@@ -190,6 +193,13 @@ set(BLAS_LIBRARIES "" CACHE STRING "The BLAS library.")
...
+
+
 # Some useful variables:
 set(CMAKE_SKIP_PREPROCESSED_SOURCE_RULES ON) # Skip *.i rules
 set(CMAKE_SKIP_ASSEMBLY_SOURCE_RULES  ON)    # Skip *.s rules
```
Comment (jandrej) [inline config/defaults.cmake:205]: extra whitespaces

File: config/defaults.mk [hunk 14]
```diff
@@ -140,6 +140,9 @@ MFEM_USE_CEED          = NO
 MFEM_USE_UMPIRE        = NO
 MFEM_USE_SIMD          = YES
 MFEM_USE_ADIOS2        = NO
+MFEM_USE_ADEPT	       = NO
```
Comment (jandrej) [inline config/defaults.mk:143]: alignment

File: examples/ex71.cpp [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+//
+// Sample runs:
+//    ex71 -m ../data/beam-quad.mesh -pp 3.5
+//    ex71 -m ../data/beam-tri.mesh  -pp 4.6
+//    ex71 -m ../data/beam-hex.mesh
+//    ex71 -m ../data/beam-tet.mesh
+//    ex71 -m ../data/beam-wedge.mesh
+//
+// Description:  This examples solves a quasi-static nonlinear
+//               pLaplacian problem with zero Dirichlet boundary
```
Comment (jandrej) [inline examples/ex71.cpp:13]: Spelling should be p-Laplacian or p-Laplace

File: .gitignore (hunk 1 above)
Comment (bslazarov) [reply .gitignore:155]: The correct number will be set just before merging as we do not know yet the last number.

Comment (jandrej) [reply .gitignore]: I was referring to the pLap directory which is created

Comment (bslazarov) [reply .gitignore]: Everything should be OK now.

**Comment by bslazarov** [comment]
> General point of view after looking at the PR for the first time.
> 
> We had an internal discussion to limit the use of templated methods. I understand the need for templated methods in this context, but this will put AD enabled builds back to square one. Just an observation.
> 
> The patch set duplicates most of the vector and matrix capabilities to introduce templated versions suitable for AD. I discourage the duplication of general functionality. This will frustrate users who use or want to try AD and want to use a feature that is available in a regular DenseMatrix but not in the templated version. Same goes for the vector and templated vector. It will also introduce a maintanance burden to keep up on both versions.

The idea is to use the same operations utilized in a standard MFEM dense vector/matrix implementation. If we change the template classes, i.e., we do not duplicate the functionality,  a user will have to implement one version for development/test proposes with AD, and later another one for production runs. In all cases, we will need a templated vector and matrix.  We can utilize an external library like Eigen - in this case; we will have one additional dependency, which is avoided with the current solution.

File: examples/ex71.cpp [hunk 16]
```diff
@@ -0,0 +1,385 @@
...
+
+      delete ns;
+      delete j_pcg;
+      delete prec;
+
+      x.SetFromTrueDofs(sv);
+      dacol->SetTime(2.0);
+      dacol->SetCycle(2);
+      dacol->Save();
+   }
```
Comment (jandrej) [inline examples/ex71.cpp:218]: This region is executed multiple times in the same way. Please make this a function and call it from the loops or special cases.

File: examples/ex71.hpp [hunk 17]
```diff
@@ -0,0 +1,562 @@
...
+   }
+};
+
+typedef ADQFunctionTH<MyQFunctorH> pLapIntegrandTH;
+
+//comment the line below in order to use
+//pLapIntegrandTJ for differentiation
+//the user interface for both TH and TJ versions
+//is exactly the same
+//#define USE_ADH
```
Comment (jandrej) [inline examples/ex71.hpp:73]: Make this a runtime selectable option. It can be compiled no matter what but should be selectable from the command line without changing source files. Then we can also properly add this to the tests.

File: examples/ex71.cpp [hunk 18]
```diff
@@ -0,0 +1,385 @@
...
+//           Hessian) are evaluated using AD.
+//
+//           We recommend viewing examples 1 and 19, before viewing this
+//           example.
+
+#include "ex71.hpp"
+
+using namespace mfem;
+
+#undef MFEM_USE_SUITESPARSE
```
Comment (jandrej) [inline examples/ex71.cpp:40]: Why is this here?

File: examples/ex71.cpp [hunk 19]
```diff
@@ -0,0 +1,385 @@
...
+   const char *mesh_file = "../data/beam-tet.mesh";
+   int ser_ref_levels = 3;
+   int order = 1;
+   bool visualization = true;
+   double newton_rel_tol = 1e-4;
+   double newton_abs_tol = 1e-6;
+   int newton_iter = 500;
+   int print_level = 0;
+   double pp = 2.0;
+   int integrator = 1; // use AD
```
Comment (jandrej) [inline examples/ex71.cpp:54]: Variable naming and/or comment unclear

File: examples/ex71.cpp [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+   // 1. Parse command-line options
+   const char *mesh_file = "../data/beam-tet.mesh";
+   int ser_ref_levels = 3;
+   int order = 1;
+   bool visualization = true;
+   double newton_rel_tol = 1e-4;
+   double newton_abs_tol = 1e-6;
+   int newton_iter = 500;
+   int print_level = 0;
+   double pp = 2.0;
```
Comment (jandrej) [inline examples/ex71.cpp:53]: Variable naming and/or comment unclear

File: examples/ex71.cpp [hunk 21]
```diff
@@ -0,0 +1,385 @@
...
+   for (int lev = 0; lev < ser_ref_levels; lev++)
+   {
+      mesh->UniformRefinement();
+   }
+
+   // 4. Define the power parameter for the p-Laplacian and all other
+   //    coefficients
+   ConstantCoefficient c_pp(pp);
+   ConstantCoefficient load(1.000000000);
+   ConstantCoefficient c_ee(0.000000001);
```
Comment (jandrej) [inline examples/ex71.cpp:119]: Are `1.0` and `1e-9` sufficient?

File: examples/ex71.hpp [hunk 22]
```diff
@@ -0,0 +1,562 @@
...
+                                   ElementTransformation &trans,
+                                   const Vector &elfun) override
+   {
+      double energy = 0.0;
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      const IntegrationRule *ir = NULL;
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
```
Comment (jandrej) [inline examples/ex71.hpp:364]: Is it the correct order? Check and remove comment.

File: examples/ex71.hpp [hunk 23]
```diff
@@ -0,0 +1,562 @@
...
+   virtual void AssembleElementGrad(const FiniteElement &el,
+                                    ElementTransformation &trans,
+                                    const Vector &elfun,
+                                    DenseMatrix &elmat) override
+   {
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      const IntegrationRule *ir = NULL;
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
```
Comment (jandrej) [inline examples/ex71.hpp:262]: Is it the correct order? Check and remove comment.

File: examples/ex71.hpp [hunk 24]
```diff
@@ -0,0 +1,562 @@
...
+                                   ElementTransformation &trans,
+                                   const Vector &elfun) override
+   {
+      double energy = 0.0;
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      const IntegrationRule *ir = NULL;
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
```
Comment (jandrej) [inline examples/ex71.hpp:111]: Is it the correct order? Check and remove comment.

File: examples/ex71.hpp [hunk 25]
```diff
@@ -0,0 +1,562 @@
...
+                                      ElementTransformation &trans,
+                                      const Vector &elfun,
+                                      Vector &elvect) override
+   {
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      const IntegrationRule *ir = NULL;
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
```
Comment (jandrej) [inline examples/ex71.hpp:429]: Is it the correct order? Check and remove comment.

File: fem/adnonlininteg.hpp [hunk 26]
```diff
@@ -0,0 +1,384 @@
...
+#if defined MFEM_USE_ADEPT
+#include <adept.h>
+#elif defined MFEM_USE_FADBADPP
+#include <fadiff.h>
+#include <badiff.h>
+#endif
+
+//define Forward AD mode
+//#define MFEM_USE_ADFORWARD
+
```
Comment (jandrej) [inline fem/adnonlininteg.hpp:33]: This should be removed

File: fem/adnonlininteg.hpp [hunk 27]
```diff
@@ -0,0 +1,384 @@
...
+namespace mfem
+{
+// m - dimension of the residual vector
+// the Jacobian will have dimensions [m,length(uu)]
+template<template<typename, typename> class CTD, int m>
+class ADQFunctionTJ
```
Comment (jandrej) [inline fem/adnonlininteg.hpp:39]: This needs proper documentation.

File: fem/adnonlininteg.hpp [hunk 28]
```diff
@@ -0,0 +1,384 @@
...
+      }
+#endif
+   }
+};
+
+//template class for differentiation; the function
+//for differentiation is supplied as a functor
+//the operator()(scalar,vector) defines the actual function
+template<template<typename, typename> class CTD>
+class ADQFunctionTH
```
Comment (jandrej) [inline fem/adnonlininteg.hpp:255]: This needs proper documentation.

File: fem/nonlinearform.cpp [hunk 29]
```diff
@@ -581,6 +581,13 @@ double BlockNonlinearForm::GetEnergyBlocked(const BlockVector &bx) const
          }
       }
 
+   // Free the allocated memory
+   for (int i = 0; i < fes.Size(); ++i)
+   {
+      delete el_x[i];
+      delete vdofs[i];
+   }
+
```
Comment (jandrej) [inline fem/nonlinearform.cpp:590]: Is this a leak fix that is not yet in master?

File: linalg/fdual.hpp [hunk 30]
```diff
@@ -0,0 +1,512 @@
...
+#include <cmath>
+#include <type_traits>
+
+namespace mfem
+{
+namespace ad
+{
+// Forward AD implementation based on dual numbers
+template<typename tbase>
+class FDual
```
Comment (jandrej) [inline linalg/fdual.hpp:24]: Documentation

File: linalg/linalg.hpp [hunk 31]
```diff
@@ -28,6 +28,7 @@
 #include "solvers.hpp"
 #include "handle.hpp"
 #include "invariants.hpp"
+// #include "fdual.hpp"
```
Comment (jandrej) [inline linalg/linalg.hpp:31]: Can this be removed?

File: fem/nonlinearform.cpp (hunk 29 above)
Comment (bslazarov) [reply fem/nonlinearform.cpp:590]: yes - it is in separate pull request - I guess it will be merged soon in master

File: linalg/fdual.hpp [hunk 32]
```diff
@@ -0,0 +1,628 @@
...
+{
+
+namespace ad
+{
+// Forward AD - simple class for automatic differentiation
+template<typename tbase>
+class FDual
+{
+private:
+   tbase pr;
```
[the remark this answers is not in the document]
Comment (jamiebramwell) [reply linalg/fdual.hpp:27]: Between the lack of comments and short variable names, this file is very hard to read.

File: fem/adnonlininteg.hpp [hunk 33]
```diff
@@ -0,0 +1,402 @@
...
+
+#ifdef MFEM_USE_ADEPT
+   ADQFunctionTJ():m_stack(false)  {}
+#else
+   ADQFunctionTJ() {}
+#endif
+
+   ~ADQFunctionTJ() {}
+
+   double QFunction(const mfem::Vector& vparam, mfem::Vector& uu)
```
Comment (jamiebramwell) [inline fem/adnonlininteg.hpp]: I don't think I can review this file until the comments are added.

File: examples/ex71.hpp [hunk 34]
```diff
@@ -0,0 +1,587 @@
...
+#include <memory>
+#include <iostream>
+#include <fstream>
+
+
+namespace mfem
+{
+
+
+template<typename DType, typename MVType>
```
Comment (jamiebramwell) [inline examples/ex71.hpp:14]: I assume these are data types and vector types?

File: fem/adnonlininteg.hpp (hunk 28 above)
Comment (jamiebramwell) [reply fem/adnonlininteg.hpp:255]: I think it also needs a readme with how you expect a user to actually use it.

Comment (jamiebramwell) [reply fem/adnonlininteg.hpp:255]: Or at least a large comment block.

File: examples/ex71.hpp [hunk 35]
```diff
@@ -0,0 +1,587 @@
...
+class MyQFunctorJ
+{
+public:
+   DType operator()(const mfem::Vector& vparam, MVType& uu)
```
Comment (jamiebramwell) [inline examples/ex71.hpp]: The requirements for how to define these functors should be documented.

File: examples/ex71.hpp [hunk 36]
```diff
@@ -0,0 +1,587 @@
...
+typedef ADQFunctionTJ<MyQFunctorJ,4> pLapIntegrandTJ;
+
+
+template<typename DType, typename MVType>
+class MyQFunctorH
+{
+public:
+   DType operator()(const mfem::Vector& vparam, MVType& uu)
+   {
+      double pp=vparam[0];
```
Comment (jamiebramwell) [inline examples/ex71.hpp]: I think an `enum class` would be better than a special index here.

File: examples/ex71.hpp [hunk 37]
```diff
@@ -0,0 +1,587 @@
...
+      DType  norm2=uu[0]*uu[0]+uu[1]*uu[1]+uu[2]*uu[2];
+
+      DType rez= pow(ee*ee+norm2,pp/2.0)/pp-ff*u;
+      return rez;
+   }
+};
+
+typedef ADQFunctionTH<MyQFunctorH> pLapIntegrandTH;
+
+//comment the line below in order to use
```
Comment (jamiebramwell) [inline examples/ex71.hpp:69]: What do you mean by `TH` and `TJ`?

File: examples/ex71.hpp [hunk 38]
```diff
@@ -0,0 +1,587 @@
...
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
+      ir = &mfem::IntRules.Get(el.GetGeomType(), order);
+
+      mfem::Vector shapef(ndof);
+      mfem::DenseMatrix dshape_iso(ndof,ndim);
+      mfem::DenseMatrix dshape_xyz(ndof,spaceDim);
+      mfem::Vector grad(spaceDim);
+
+      mfem::Vector vparam(3);//[power, epsilon, load]
+      mfem::Vector uu(4);//[diff_x,diff_y,diff_z,u]
```
Comment (jamiebramwell) [inline examples/ex71.hpp]: Special indices should be captured in enums.

File: CMakeLists.txt [hunk 39]
```diff
@@ -326,6 +326,34 @@ if (MFEM_USE_HIOP)
...
 endif()
 
+# ADEPT package
+if (MFEM_USE_ADEPT)
+        find_package(ADEPT REQUIRED)
+        # find_package updates ADEPT_FOUND, ADEPT_INCLUDE_DIRS, ADEPT_LIBRARIES
+endif()
+
+# FADBAD++ package
+if (MFEM_USE_FADBADPP)
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply CMakeLists.txt:336]: I just documented TADVector, TADDenseMatrix, ADQFunctionTJ and ADQFunctionTH. Any comments are welcome.

File: fem/adnonlininteg.hpp (hunk 27 above)
Comment (bslazarov) [reply fem/adnonlininteg.hpp:90]: Documented!

File: fem/adnonlininteg.hpp (hunk 28 above)
Comment (bslazarov) [reply fem/adnonlininteg.hpp:361]: Documented! Please, return comments!

File: linalg/fdual.hpp (hunk 30 above)
Comment (bslazarov) [reply linalg/fdual.hpp:50]: Documented!

File: .gitignore [hunk 40]
```diff
@@ -147,6 +147,11 @@ examples/hiop/ex9-mesh.*
 examples/hiop/ex9-init.*
 examples/hiop/ex9-final.*
 
+
+examples/ex71
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply .gitignore:150]: It will be renamed before merging - I am trying to avoid name changes during the development. I am periodically merging the master branch in the current one.

File: examples/ex71.cpp (hunk 16 above)
Comment (bslazarov) [reply examples/ex71.cpp]: extracted and separated in a class

File: examples/ex71.cpp (hunk 19 above)
Comment (bslazarov) [reply examples/ex71.cpp]: New comments in the code - added

File: linalg/linalg.hpp (hunk 31 above)
Comment (bslazarov) [reply linalg/linalg.hpp:31]: removed

File: examples/ex71.hpp (hunk 17 above)
Comment (bslazarov) [reply examples/ex71.hpp:106]: it is implemented now

File: examples/ex71.cpp [hunk 41]
```diff
@@ -0,0 +1,348 @@
...
+   const char *mesh_file = "../data/beam-tet.mesh";
+   int ser_ref_levels = 3;
+   int order = 1;
+   bool visualization = true;
+   double newton_rel_tol = 1e-4;
+   double newton_abs_tol = 1e-6;
+   int newton_iter = 500;
+   int print_level = 0;
+   double pp = 2.0;
+   int integrator=1; //use AD
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply examples/ex71.cpp]: Code is commented and modified accordingly - let me know how it reads.

File: examples/ex71.cpp [hunk 42]
```diff
@@ -0,0 +1,348 @@
...
+   for (int lev = 0; lev < ser_ref_levels; lev++)
+   {
+      mesh->UniformRefinement();
+   }
+
+   // 4. Define the power parameter for the p-Laplacian and all other
+   //    coefficients
+   mfem::ConstantCoefficient c_pp(pp);
+   mfem::ConstantCoefficient load(1.000000000);
+   mfem::ConstantCoefficient c_ee(0.000000001);
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply examples/ex71.cpp]: There are separated in a class and the naming is changed

File: examples/ex71.cpp [hunk 43]
```diff
@@ -0,0 +1,348 @@
...
+      delete j_pcg;
+      delete prec;
+
+      x.SetFromTrueDofs(sv);
+      dacol->SetTime(2.0);
+      dacol->SetCycle(2);
+      dacol->Save();
+   }
+
+   // 12. Continue with powers higher than 2
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply examples/ex71.cpp]: new comments in the code :-)

File: examples/ex71.hpp (hunk 35 above)
Comment (bslazarov) [reply examples/ex71.hpp]: Documented!

File: examples/ex71.hpp (hunk 36 above)
Comment (bslazarov) [reply examples/ex71.hpp]: We do not know in advance the number of the variables - this means that we have to define enum for every separate case.

File: examples/ex71.hpp (hunk 37 above)
Comment (bslazarov) [reply examples/ex71.hpp:102]: H is for Hessian and J is for Jacobian

File: examples/ex71.hpp (hunk 38 above)
Comment (bslazarov) [reply examples/ex71.hpp]: Again - these indices are case dependent - I am not sure enum is the correct solution here as it will have to be defined for every new implementation.

File: examples/ex71.cpp [hunk 44]
```diff
@@ -0,0 +1,348 @@
...
+   mfem::Array<int> ess_bdr(mesh->bdr_attributes.Max());
+   ess_bdr = 1;
+
+   // 7. Define the nonlinear form
+   mfem::NonlinearForm* nf=new mfem::NonlinearForm(&fespace);
+
+   // 8. Define the solution vector x
+   mfem::GridFunction x(&fespace);
+   x = 0.0;
+   mfem::Vector tv(fespace.GetTrueVSize());
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply examples/ex71.cpp]: removed

File: examples/ex71.cpp [hunk 45]
```diff
@@ -0,0 +1,348 @@
...
+         nf->AddDomainIntegrator(new mfem::pLaplace(lpp,c_ee,load));
+      }
+      else if (integrator==1)
+      {
+         nf->AddDomainIntegrator(new mfem::pLaplaceAD(lpp,c_ee,load));
+      }
+      nf->SetEssentialBC(ess_bdr);
+      // compute the energy
+      double energy=nf->GetEnergy(tv);
+      std::cout<<"[2] The total energy of the system is E="<<energy<<std::endl;
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply examples/ex71.cpp]: I do not see anything wrong with it?

File: linalg/tadvector.hpp [hunk 46]
```diff
@@ -0,0 +1,703 @@
...
+                  "index [" << i << "] is out of range [0," << size << ")");
+
+      return data[i];
+   }
+
+   /// Dot product with a `dtype *` array.
+   dtype operator*(const dtype *v) const
+   {
+      dtype dot = 0.0;
+#ifdef MFEM_USE_LEGACY_OPENMP
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply linalg/tadvector.hpp:321]: removed/

File: linalg/tadvector.hpp [hunk 47]
```diff
@@ -0,0 +1,703 @@
...
+      SetSize(v.Size());
+      for (int i=0; i<size; i++)
+      {
+         data[i]=v[i];
+      }
+      return *this;
+   }
+
+   /// Redefine '=' for vector = constant.
+   template<typename ivtype>
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply linalg/tadvector.hpp:388]: they could be double and float for example - templates will automatically define the method for different types. I can make them to be only for the base type - let me know if you think that it will be better. 

I do not think that MFEM has style guide for these cases.

File: linalg/tadvector.hpp [hunk 48]
```diff
@@ -0,0 +1,703 @@
...
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_TADVECTOR
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply linalg/tadvector.hpp:12]: We had many discussions - templates would require significant changes for the standard vector class - therefore, we decided to separate them for now.

File: linalg/fdual.hpp (hunk 32 above)
[the remark this answers is not in the document]
Comment (bslazarov) [reply linalg/fdual.hpp:54]: The class is documented.

File: fem/adnonlininteg.hpp (hunk 33 above)
Comment (bslazarov) [reply fem/adnonlininteg.hpp]: Comments have been added in the code.

File: examples/ex71.hpp (hunk 34 above)
Comment (bslazarov) [reply examples/ex71.hpp:20]: Yes they are. The new code is commented.

File: examples/ex71.cpp (hunk 21 above)
Comment (bslazarov) [reply examples/ex71.cpp]: They are just parameters - they will define the equation which is solved.

File: examples/ex71.hpp (hunk 22 above)
Comment (bslazarov) [reply examples/ex71.hpp]: corrected

File: fem/adnonlininteg.hpp (hunk 26 above)
Comment (bslazarov) [reply fem/adnonlininteg.hpp:33]: removed

File: examples/ex71.cpp (hunk 20 above)
Comment (bslazarov) [reply examples/ex71.cpp]: comments have been added.

**Comment by tzanio** [comment]
This work will be refactored in a new PR
